File: src/main/java/me/miki/shindo/management/mods/impl/crosshair/LayoutManager.java
package me.miki.shindo.management.mods.impl.crosshair;

import com.google.gson.*;
import me.miki.shindo.Shindo;
import me.miki.shindo.management.file.FileManager;
import me.miki.shindo.utils.JsonUtils;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class LayoutManager {

    public static final int MAX_CUSTOM_PRESETS = 8;
    private static final int DEFAULT_PRESET_COLOR = 0xFFFF0000;
    private static final String PRESET_FILE_NAME = "CrosshairPresets.json";

    private final boolean[][] preset4 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset5 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, true, true, true, false, false, false, false},
            {false, false, false, true, true, false, true, true, false, false, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, false, false, true, true, false, true, true, false, false, false},
            {false, false, false, false, true, true, true, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset7 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, false, false, false, true, false, true, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, true, false, true, false, false, false, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset8 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, true, false, true, false, true, false, false, false},
            {false, false, true, false, false, true, false, false, true, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, true, true, true, false, true, false, true, true, true, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, true, false, false, true, false, false, true, false, false},
            {false, false, false, true, false, true, false, true, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset11 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, true, true, true, false, false, false, false},
            {false, false, false, true, true, false, true, true, false, false, false},
            {false, false, true, true, false, true, false, true, true, false, false},
            {false, false, false, true, true, false, true, true, false, false, false},
            {false, false, false, false, true, true, true, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset13 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, true, true, false, false, false, true, true, false, false},
            {false, false, true, false, false, false, false, false, true, false, false},
            {false, false, true, false, false, true, false, false, true, false, false},
            {false, false, true, false, false, false, false, false, true, false, false},
            {false, false, true, true, false, false, false, true, true, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset14 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, true, true, true, true, true, false, false, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, false, false, true, true, true, true, true, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset15 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, true, false, true, false, false, false, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, true, true, false, false, true, false, false, true, true, false},
            {false, false, false, true, false, false, false, true, false, false, false},
            {false, false, false, false, true, false, true, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };
    private final boolean[][] preset16 = new boolean[][]{
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, true, true, true, false, false, false, false},
            {false, false, false, false, false, true, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false},
            {false, false, false, false, false, false, false, false, false, false, false}
    };

    private final List<CellGridPreset> userPresets = new ArrayList<>();
    private final File presetFile;

    public LayoutManager() {
        FileManager fileManager = Shindo.getInstance().getFileManager();
        this.presetFile = new File(fileManager.getShindoDir(), PRESET_FILE_NAME);
        init();
    }

    public void init() {
        userPresets.clear();
        loadFromDisk();

        if (userPresets.isEmpty()) {
            seedDefaults();
        }
    }

    public boolean[][] getDefaultLayout() {
        return copyGrid(preset4);
    }

    public List<CellGridPreset> getCustomPresets() {
        return new ArrayList<>(userPresets);
    }

    public CellGridPreset addCustomPreset(String name, boolean[][] layout, int[][] colors) {
        return addOrUpdatePreset(null, layout, colors, name);
    }

    public CellGridPreset addOrUpdatePreset(String id, boolean[][] layout, int[][] colors, String nameOverride) {
        CellGridPreset target = id == null ? null : getPresetById(id);

        if (target != null) {
            target.update(layout, colors);
            saveToDisk();
            return target;
        }

        CellGridPreset preset = new CellGridPreset(
                id == null ? UUID.randomUUID().toString() : id,
                nameOverride == null ? "" : nameOverride,
                layout,
                colors,
                true
        );
        userPresets.removeIf(p -> p.getId().equals(preset.getId()));
        userPresets.add(preset);
        enforceCustomLimit();
        saveToDisk();
        return preset;
    }

    public CellGridPreset getPresetById(String id) {
        if (id == null) {
            return null;
        }
        for (CellGridPreset preset : userPresets) {
            if (id.equals(preset.getId())) {
                return preset;
            }
        }
        return null;
    }

    public void removePreset(CellGridPreset preset) {
        if (preset == null) {
            return;
        }
        userPresets.removeIf(entry -> entry == preset || entry.getId().equals(preset.getId()));
        saveToDisk();
    }

    private void enforceCustomLimit() {
        while (userPresets.size() > MAX_CUSTOM_PRESETS) {
            userPresets.remove(0);
        }
    }

    private void seedDefaults() {
        addOrUpdatePreset("seed-dot", preset4, null, "dot");
        addOrUpdatePreset("seed-diamond", preset5, null, "diamond");
        addOrUpdatePreset("seed-star", preset8, null, "star");
    }

    private void loadFromDisk() {
        try {
            if (!presetFile.exists()) {
                presetFile.createNewFile();
                return;
            }

            try (FileReader reader = new FileReader(presetFile)) {
                Gson gson = new GsonBuilder().setPrettyPrinting().create();
                JsonObject json = gson.fromJson(reader, JsonObject.class);
                if (json == null || !json.has("presets")) {
                    return;
                }

                JsonArray array = json.getAsJsonArray("presets");
                Iterator<JsonElement> iterator = array.iterator();
                while (iterator.hasNext() && userPresets.size() < MAX_CUSTOM_PRESETS) {
                    JsonObject element = iterator.next().getAsJsonObject();
                    boolean[][] layout = JsonUtils.parseBooleanGrid(element.get("layout"));
                    int[][] colors = JsonUtils.parseIntGrid(element.get("colors"));
                    String id = JsonUtils.getStringProperty(element, "id", UUID.randomUUID().toString());
                    String name = JsonUtils.getStringProperty(element, "name", "");
                    if (layout != null) {
                        userPresets.add(new CellGridPreset(id, name, layout, colors, true));
                    }
                }
            }
        } catch (Exception ignored) {
        }
    }

    public void saveToDisk() {
        try (FileWriter writer = new FileWriter(presetFile)) {
            Gson gson = new GsonBuilder().setPrettyPrinting().create();
            JsonObject root = new JsonObject();
            JsonArray presetsArray = new JsonArray();

            for (CellGridPreset preset : userPresets) {
                JsonObject entry = new JsonObject();
                entry.addProperty("id", preset.getId());
                entry.addProperty("name", preset.getName());
                entry.add("layout", JsonUtils.toBooleanGrid(preset.getLayoutCopy()));
                entry.add("colors", JsonUtils.toIntGrid(preset.getColorCopy()));
                presetsArray.add(entry);
            }

            root.add("presets", presetsArray);
            gson.toJson(root, writer);
        } catch (Exception ignored) {
        }
    }

    private static boolean[][] copyGrid(boolean[][] source) {
        if (source == null) {
            return null;
        }
        boolean[][] copy = new boolean[source.length][];
        for (int i = 0; i < source.length; i++) {
            boolean[] row = source[i];
            copy[i] = row != null ? row.clone() : null;
        }
        return copy;
    }

    private static int[][] copyColors(int[][] source, boolean[][] layout) {
        if (layout == null) {
            return null;
        }
        int[][] copy = new int[layout.length][];
        for (int i = 0; i < layout.length; i++) {
            int length = layout[i] != null ? layout[i].length : 0;
            copy[i] = new int[length];
            if (source != null && i < source.length && source[i] != null) {
                System.arraycopy(source[i], 0, copy[i], 0, Math.min(length, source[i].length));
            }
            for (int j = 0; j < length; j++) {
                if (copy[i][j] == 0) {
                    copy[i][j] = DEFAULT_PRESET_COLOR;
                }
            }
        }
        return copy;
    }

    public static final class CellGridPreset {
        private final String id;
        private final String name;
        private boolean[][] layout;
        private int[][] colors;
        private final boolean userPreset;

        private CellGridPreset(String id, String name, boolean[][] layout, int[][] colors, boolean userPreset) {
            this.id = id;
            this.name = name;
            this.layout = copyGrid(layout);
            this.colors = copyColors(colors, layout);
            this.userPreset = userPreset;
        }

        public String getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public boolean[][] getLayoutCopy() {
            return copyGrid(layout);
        }

        public int[][] getColorCopy() {
            return copyColors(colors, layout);
        }

        public void update(boolean[][] newLayout, int[][] newColors) {
            this.layout = copyGrid(newLayout);
            this.colors = copyColors(newColors, newLayout);
        }

        public boolean isUserPreset() {
            return userPreset;
        }
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/crosshair/LayoutManager.kt
package me.miki.shindo.management.mods.impl.crosshair

import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import me.miki.shindo.Shindo
import me.miki.shindo.utils.JsonUtils
import me.miki.shindo.utils.JsonUtils.getStringProperty
import me.miki.shindo.utils.JsonUtils.parseBooleanGrid
import me.miki.shindo.utils.JsonUtils.parseIntGrid
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.util.*
import kotlin.math.min

@Suppress("UNUSED")
class LayoutManager {
    private val preset4 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset5 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, true, true, true, false, false, false, false),
            booleanArrayOf(false, false, false, true, true, false, true, true, false, false, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, false, false, true, true, false, true, true, false, false, false),
            booleanArrayOf(false, false, false, false, true, true, true, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset7 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, false, false, false, true, false, true, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, true, false, true, false, false, false, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset8 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, true, false, true, false, true, false, false, false),
            booleanArrayOf(false, false, true, false, false, true, false, false, true, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, true, true, true, false, true, false, true, true, true, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, true, false, false, true, false, false, true, false, false),
            booleanArrayOf(false, false, false, true, false, true, false, true, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset11 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, true, true, true, false, false, false, false),
            booleanArrayOf(false, false, false, true, true, false, true, true, false, false, false),
            booleanArrayOf(false, false, true, true, false, true, false, true, true, false, false),
            booleanArrayOf(false, false, false, true, true, false, true, true, false, false, false),
            booleanArrayOf(false, false, false, false, true, true, true, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset13 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, true, true, false, false, false, true, true, false, false),
            booleanArrayOf(false, false, true, false, false, false, false, false, true, false, false),
            booleanArrayOf(false, false, true, false, false, true, false, false, true, false, false),
            booleanArrayOf(false, false, true, false, false, false, false, false, true, false, false),
            booleanArrayOf(false, false, true, true, false, false, false, true, true, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset14 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, true, true, true, true, true, false, false, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, false, false, true, true, true, true, true, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset15 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, true, false, true, false, false, false, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, true, true, false, false, true, false, false, true, true, false),
            booleanArrayOf(false, false, false, true, false, false, false, true, false, false, false),
            booleanArrayOf(false, false, false, false, true, false, true, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )
    private val preset16 =
        arrayOf<BooleanArray?>(
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, true, true, true, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, true, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
            booleanArrayOf(false, false, false, false, false, false, false, false, false, false, false),
        )

    private val userPresets: MutableList<CellGridPreset> = ArrayList<CellGridPreset>()
    private val presetFile: File

    init {
        val fileManager = Shindo.getInstance().getFileManager()
        this.presetFile = File(fileManager.shindoDir, PRESET_FILE_NAME)
        init()
    }

    fun init() {
        userPresets.clear()
        loadFromDisk()

        if (userPresets.isEmpty()) {
            seedDefaults()
        }
    }

    val defaultLayout: Array<BooleanArray?>
        get() = copyGrid(preset4)

    val customPresets: MutableList<CellGridPreset?>
        get() = ArrayList<CellGridPreset?>(userPresets)

    fun addCustomPreset(
        name: String?,
        layout: Array<BooleanArray?>,
        colors: Array<IntArray?>,
    ): CellGridPreset = addOrUpdatePreset(null, layout, colors, name)

    fun addOrUpdatePreset(
        id: String?,
        layout: Array<BooleanArray?>,
        colors: Array<IntArray?>,
        nameOverride: String?,
    ): CellGridPreset {
        val target = if (id == null) null else getPresetById(id)

        if (target != null) {
            target.update(layout, colors)
            saveToDisk()
            return target
        }

        val preset =
            CellGridPreset(
                id ?: UUID.randomUUID().toString(),
                nameOverride ?: "",
                layout,
                colors,
                true,
            )
        userPresets.removeIf { p: CellGridPreset? -> p!!.id == preset.id }
        userPresets.add(preset)
        enforceCustomLimit()
        saveToDisk()
        return preset
    }

    fun getPresetById(id: String?): CellGridPreset? {
        if (id == null) {
            return null
        }
        for (preset in userPresets) {
            if (id == preset.id) {
                return preset
            }
        }
        return null
    }

    fun removePreset(preset: CellGridPreset?) {
        if (preset == null) {
            return
        }
        userPresets.removeIf { entry: CellGridPreset? -> entry == preset || entry!!.id == preset.id }
        saveToDisk()
    }

    private fun enforceCustomLimit() {
        while (userPresets.size > MAX_CUSTOM_PRESETS) {
            userPresets.removeAt(0)
        }
    }

    private fun seedDefaults() {
        addOrUpdatePreset("seed-dot", preset4, arrayOfNulls(0), "dot")
        addOrUpdatePreset("seed-diamond", preset5, arrayOfNulls(0), "diamond")
        addOrUpdatePreset("seed-star", preset8, arrayOfNulls(0), "star")
    }

    private fun loadFromDisk() {
        try {
            if (!presetFile.exists()) {
                presetFile.createNewFile()
                return
            }

            FileReader(presetFile).use { reader ->
                val gson = GsonBuilder().setPrettyPrinting().create()
                val json = gson.fromJson(reader, JsonObject::class.java)
                if (json == null || !json.has("presets")) {
                    return
                }

                val array = json.getAsJsonArray("presets")
                val iterator = array.iterator()
                while (iterator.hasNext() && userPresets.size < MAX_CUSTOM_PRESETS) {
                    val element = iterator.next()!!.asJsonObject
                    val layout: Array<BooleanArray?> = parseBooleanGrid(element.get("layout"))
                    val colors: Array<IntArray?> = parseIntGrid(element.get("colors"))
                    val id = getStringProperty(element, "id", UUID.randomUUID().toString())
                    val name = getStringProperty(element, "name", "")
                    userPresets.add(CellGridPreset(id, name, layout, colors, true))
                }
            }
        } catch (ignored: Exception) {
        }
    }

    fun saveToDisk() {
        try {
            FileWriter(presetFile).use { writer ->
                val gson = GsonBuilder().setPrettyPrinting().create()
                val root = JsonObject()
                val presetsArray = JsonArray()

                for (preset in userPresets) {
                    val entry = JsonObject()
                    entry.addProperty("id", preset.id)
                    entry.addProperty("name", preset.name)
                    entry.add("layout", JsonUtils.toBooleanGrid(preset.layoutCopy))
                    entry.add("colors", JsonUtils.toIntGrid(preset.colorCopy))
                    presetsArray.add(entry)
                }

                root.add("presets", presetsArray)
                gson.toJson(root, writer)
            }
        } catch (ignored: Exception) {
        }
    }

    class CellGridPreset(
        val id: String?,
        val name: String?,
        layout: Array<BooleanArray?>,
        colors: Array<IntArray?>,
        val isUserPreset: Boolean,
    ) {
        private var layout: Array<BooleanArray?>
        private var colors: Array<IntArray?>

        init {
            this.layout = copyGrid(layout)
            this.colors = copyColors(colors, layout)
        }

        val layoutCopy: Array<BooleanArray?>
            get() = copyGrid(layout)

        val colorCopy: Array<IntArray?>
            get() = copyColors(colors, layout)

        fun update(
            newLayout: Array<BooleanArray?>,
            newColors: Array<IntArray?>,
        ) {
            this.layout = copyGrid(newLayout)
            this.colors = copyColors(newColors, newLayout)
        }
    }

    companion object {
        const val MAX_CUSTOM_PRESETS: Int = 8
        private const val DEFAULT_PRESET_COLOR = -0x10000
        private const val PRESET_FILE_NAME = "CrosshairPresets.json"

        private fun copyGrid(source: Array<BooleanArray?>): Array<BooleanArray?> {
            val copy = arrayOfNulls<BooleanArray>(source.size)
            for (i in source.indices) {
                val row = source[i]
                copy[i] = row!!.clone()
            }
            return copy
        }

        private fun copyColors(
            source: Array<IntArray?>,
            layout: Array<BooleanArray?>,
        ): Array<IntArray?> {
            val copy = arrayOfNulls<IntArray>(layout.size)
            for (i in layout.indices) {
                val length = layout[i]!!.size
                copy[i] = IntArray(length)
                if (i < source.size) {
                    System.arraycopy(source[i]!!, 0, copy[i]!!, 0, min(length, source[i]!!.size))
                }
                for (j in 0 until length) {
                    if (copy[i]!![j] == 0) {
                        copy[i]!![j] = DEFAULT_PRESET_COLOR
                    }
                }
            }
            return copy
        }
    }
}
